lib/dao.js

```diff
@@ -624,7 +624,193 @@ DataAccessObject.upsert = function(data, options, cb) {
+          if (!(data instanceof Model)) {
+            inst = new Model(data, { applyDefaultValues: false });
+          }
+          update = inst.toObject(false);
+          Model.applyProperties(update, inst);
+          Model = Model.lookupModel(update);
+          var connector = self.getConnector();
+          var doValidate = undefined;
+          if (options.validate === undefined) {
+            if (Model.settings.automaticValidation !== undefined) {
+              doValidate = Model.settings.automaticValidation;
+            } else {
+              doValidate = Model.settings.validateUpsert;
+            }
+          } else {
+            doValidate = options.validate;
+          }
+          if (doValidate === false) {
+            callConnector();
+          } else {
+            inst.isValid(function(valid) {
+              if (!valid) {
+                if (doValidate) {
+                  return cb(new ValidationError(inst), inst);
+                } else {
+                  console.warn('Ignoring validation errors in updateOrCreate():');
+                  console.warn('  %s', new ValidationError(inst).message);
+                }
+              }
+              callConnector();
+            }, update, options);
+          }
+
+          function callConnector() {
+            update = removeUndefined(update);
+            context = {
+              Model: Model,
+              where: ctx.where,
+              data: update,
+              currentInstance: inst,
+              hookState: ctx.hookState,
+              options: options,
+            };
+            Model.notifyObserversOf('persist', context, function(err) {
+              if (err) return done(err);
+              connector.upsertWithWhere(modelName, where, update, options, done);
+            });
+          }
+          function done(err, data, info) {
+            if (err) return cb(err);
+            var context = {
+              Model: Model,
+              data: data,
+              isNewInstance: info && info.isNewInstance,
+              hookState: ctx.hookState,
+              options: options,
+            };
+            Model.notifyObserversOf('loaded', context, function(err) {
+              if (err) return cb(err);
+              var obj;
+              if (data && !(data instanceof Model)) {
+                inst._initProperties(data, { persisted: true });
+                obj = inst;
+              } else {
+                obj = data;
+              }
 
+              var context = {
+                Model: Model,
+                instance: obj,
+                isNewInstance: info ? info.isNewInstance : undefined,
+                hookState: hookState,
+                options: options,
+              };
+              Model.notifyObserversOf('after save', context, function(err) {
+                cb(err, obj);
+              });
+            });
+          }
+        });
+      } else {
+        var element = Object.keys(where)[0];
```

Comment:
I have mentioned this issue in multiple comments and they have not been addressed; please see [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/1001#discussion_r71026493) and [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/1001#discussion_r71170174)

We support `AND` and `OR` conditions in `where` clause as explained elsewhere and your logic is not going to work for them; Example:

```
    where: {
      or: {
        'name': {like: 'foo'},
        'familyName': {like: 'bar'}
      }
    }
```